### PR TITLE
[BEAM-9562] Add Timer to Elements proto representation.

### PR DESCRIPTION
**Pull request by boyuanzz**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 1]
```diff
@@ -164,6 +164,11 @@ message ProcessBundleDescriptor {
   // data plane or if any of the transforms rely on user state or side inputs.
   org.apache.beam.model.pipeline.v1.ApiServiceDescriptor
       state_api_service_descriptor = 7;
+
+  // A descriptor describing the end  point to use for Data API to use user
+  // timers.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
  // A descriptor describing the end point to use for Data API for user timers.
  // Required if the ProcessBundleDescriptor contains any transforms that have user timers.
```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 2]
```diff
@@ -533,8 +538,33 @@ message Elements {
     bytes data = 3;
   }
 
+  // Represent the encoded user timer for a given instruction, transform and
+  // timer id.
+  message Timer {
+    // (Required) A reference to an active instruction request with the given
+    // instruction id.
+    string instruction_id = 1;
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
    string instruction_id = 1;

```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 3]
```diff
@@ -533,8 +538,33 @@ message Elements {
...
+    // (Required) A definition representing a consumer or producer of this data.
+    // If received by a harness, this represents the consumer within that
+    // harness that should consume these bytes. If sent by a harness, this
+    // represents the producer of these bytes.
+    string transform_id = 2;
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
    string transform_id = 2;

```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 4]
```diff
@@ -533,8 +538,33 @@ message Elements {
...
+    // (Optional) The local timer name which can be mapped to
+    // timer specification.
+    string timer_id = 3;
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
    string timer_id = 3;

```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 5]
```diff
@@ -533,8 +538,33 @@ message Elements {
...
+  // Represent the encoded user timer for a given instruction, transform and
+  // timer id.
+  message Timer {
+    // (Required) A reference to an active instruction request with the given
+    // instruction id.
+    string instruction_id = 1;
+    // (Required) A definition representing a consumer or producer of this data.
+    // If received by a harness, this represents the consumer within that
+    // harness that should consume these bytes. If sent by a harness, this
+    // represents the producer of these bytes.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
    // harness that should consume these timers. If sent by a harness, this
    // represents the producer of these timers.
```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 6]
```diff
@@ -533,8 +538,33 @@ message Elements {
...
+    string transform_id = 2;
+    // (Optional) The local timer name which can be mapped to
+    // timer specification.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
    // (Optional) The local timer name used to identify the associated timer specification.
```

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 7]
```diff
@@ -533,8 +538,33 @@ message Elements {
...
+    string timer_id = 3;
+    // (Optional) Represents a logical byte stream of a timer. Encoded according
+    // to the coder in the timer spec.
+    // An empty data block represents the end of stream for the given
+    // instruction and transform.
+    bytes timer = 4;
+  }
+
+
   // (Required) A list containing parts of logical byte streams.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto]:
```suggestion
  // (Optional) A list containing parts of logical byte streams.
```

**Comment by boyuanzz** [comment]
Thanks for your review! 
Java PreCommit failures are not related. I'm going to merge it.

**Pull request merged**